### PR TITLE
yield before wait/reserve and support for matrix multiplication

### PR DESCRIPTION
**Pull request by FConstantinos**
- Allows sequential execution of compute/dm threads in round robin
- inserts a yield before blocking operations for RR scheduler to determine whether the thread can continue
- Add support for matrix multiplication

### CODE REVIEW

File: test/sim/test_program.py [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+        out = torch.zeros(TILE_SHAPE)
+
+        test_kernel(a, out)
+
+        # Verify computation
+        expected = torch.ones(TILE_SHAPE) * 6
+        tt_testing.assert_close(out, expected)
+
+    def test_modes_produce_same_result(self) -> None:
+        """Test that THREADED mode works (COOPERATIVE skipped due to getsource issues)."""
```
Comment (zoecarver) [inline test/sim/test_program.py]: What's the getsource issue?

**Review by zoecarver — APPROVED** [review]
Do we have some simulator documentation? Might be nice to document the new execution modes outside of the code. This looks good to me!

File: examples/singlecore_matmul.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        for m in range(Mt):
+            for n in range(Nt):
+                # Reserve output tile once and accumulate K contributions.
+                out_block = out_cb.reserve()  # blocking
+                for k in range(Kt):
+                    a_block = a_cb.wait()  # blocking
+                    b_block = b_cb.wait()  # blocking
+
+                    if k == 0:
+                        # Initialize output tile.
```
[the remark this answers is not in the document]
Comment (FConstantinos) [reply examples/singlecore_matmul.py]: This turns out to be a rather involved change and I'd prefer to fix it in another PR. In the meantime I've created an issue for this:

https://github.com/tenstorrent/tt-lang/issues/151
